src/ws/cockpithandlers.c

```diff
@@ -333,12 +336,35 @@ send_login_html (CockpitWebResponse *response,
   filter2 = cockpit_web_inject_new (marker, url_bytes, 1);
   g_bytes_unref (url_bytes);
   cockpit_web_response_add_filter (response, filter2);
+  g_object_unref (filter2);
 
   cockpit_web_response_set_cache_type (response, COCKPIT_WEB_RESPONSE_NO_CACHE);
-  cockpit_web_response_file (response, "/login.html", ws->static_roots);
-  g_object_unref (filter);
-  if (filter2)
-    g_object_unref (filter2);
+
+  bytes = cockpit_web_response_negotiation (ws->login_html, NULL, NULL, NULL, &error);
+  if (error)
+    {
+      if (error)
+        {
+          g_message ("%s", error->message);
+          cockpit_web_response_error (response, 500, NULL, NULL);
+        }
+    }
+  else if (!bytes)
+    {
+      cockpit_web_response_error (response, 404, NULL, NULL);
+    }
+  else
+    {
+      /* The login Content-Security-Policy allows the page to have inline <script> and <style> tags. */
+      cockpit_web_response_headers (response, 200, "OK", -1,
+                                    "Content-Security-Policy",
+                                    "default-src 'self' 'unsafe-inline'; connect-src 'self' ws: wss:",
```

Comment:
Yup. This makes sense ... we're no longer using cockpit_web_response_file() which would otherwise have done this.